### PR TITLE
Rewrite inventory Component

### PR DESCRIPTION
**Pull request by nimeshgarg**
Inventory Component rewritten to work on Item Names.

### CODE REVIEW

File: source/core/src/main/com/csse3200/game/components/player/InventoryComponent.java [hunk 1]
```diff
@@ -1,183 +1,371 @@
...
+   * @return boolean representing if the item is on the character
+   */
+  public Boolean hasItem(String item) {
+    return this.itemCount.containsKey(item);
+  }
+
+  private void newInventory() {
+    this.itemPlace = new HashMap<>();
+    this.itemCount = new HashMap<>();
+    //this.inventory = new ArrayList<>();
```
Comment (JTrenerry) [inline source/core/src/main/com/csse3200/game/components/player/InventoryComponent.java:165]: remove?

File: source/core/src/main/com/csse3200/game/components/player/InventoryComponent.java [hunk 2]
```diff
@@ -1,183 +1,371 @@
...
-   * Returns if the player has a certain amount of gold.
-   * @param item item to be checked
+   * Returns if the player has a certain item or not.
+   * @param item Entity to be checked
    * @return boolean representing if the item is on the character
    */
   public Boolean hasItem(Entity item) {
-    return this.inventory.contains(item);
+    if(item.getType() != EntityType.Item) {
+      //logger.info("Entity is not an item");
```
Comment (JTrenerry) [inline source/core/src/main/com/csse3200/game/components/player/InventoryComponent.java:147]: remove / uncomment

File: source/core/src/main/com/csse3200/game/components/player/InventoryComponent.java [hunk 3]
```diff
@@ -1,183 +1,371 @@
...
+
   public void setInventory(List<Entity> items) {
-    this.inventory.clear();
-    this.inventory.addAll(items);
-    int pos = itemPlace.size();
-    //int pos = 0;
-    logger.debug("Setting inventory to {}", this.inventory.toString());
+    newInventory();
+    //logger.debug("Setting inventory");
+    //this.inventory.addAll(items);
```
Comment (JTrenerry) [inline source/core/src/main/com/csse3200/game/components/player/InventoryComponent.java:188]: to remove?

File: source/core/src/main/com/csse3200/game/components/player/InventoryComponent.java [hunk 4]
```diff
@@ -1,183 +1,371 @@
...
-    return true;
+  /**
+   * Function to get the exchange the position of two specific item in Inventory.
+   *
+   * @param pos1 position of first Inventory Item
+   * @param pos2 position of second Inventory Item
+   */
+  public boolean swapPosition(int pos1, int pos2) {
+    if (pos1 >= this.getInventorySize() || pos1 < 0 || pos2 >= this.getInventorySize() || pos2 < 0) {
+      //logger.info("Position is out of bounds");
```
Comment (JTrenerry) [inline source/core/src/main/com/csse3200/game/components/player/InventoryComponent.java:265]: uncomment / remove

File: source/core/src/main/com/csse3200/game/components/player/InventoryComponent.java [hunk 5]
```diff
@@ -1,183 +1,371 @@
...
+
+  public boolean setPosition(Entity entity, int position) {
+    if (this.itemPlace.containsValue(entity.getComponent(ItemComponent.class).getItemName())) {
+      //logger.info("Item already in inventory, Swap position instead");
+      return false;
+    } else if (position >= this.getInventorySize() || position < 0) {
+      //logger.info("Position is out of bounds");
+      return false;
+    } else if (this.itemPlace.get(position) != null) {
+      //logger.info("Position is already occupied");
```
Comment (JTrenerry) [inline source/core/src/main/com/csse3200/game/components/player/InventoryComponent.java:293]: uncomment / remove

File: source/core/src/main/com/csse3200/game/components/player/InventoryComponent.java [hunk 6]
```diff
@@ -1,183 +1,371 @@
...
-    int lastPlace = findFirstIndex();
-    if (lastPlace == -1) {
+    int position = nextAvailablePosition();
+    if (position == -1) {
       return false;
+    } else if (this.itemPlace.get(position) != null) {
+        //logger.info("Position is already occupied");
+        return false;
+    } else if (this.itemPlace.containsValue(entity.getComponent(ItemComponent.class).getItemName())) {
+        //logger.info("Item already in inventory, Swap position instead");
```
Comment (JTrenerry) [inline source/core/src/main/com/csse3200/game/components/player/InventoryComponent.java:316]: uncomment / remove

File: source/core/src/main/com/csse3200/game/components/player/InventoryComponent.java [hunk 7]
```diff
@@ -1,183 +1,371 @@
...
-      itemPosition.put(item, new Point(0, 0)); // Default position. You can change this as needed.
+    if(item.getType() != EntityType.Item || item.getComponent(ItemComponent.class) == null) {
+      //logger.info("Entity is not an item");
+      //logger.info(entity.toString());
+      return false;
+    }
+    if (isFull()) {
+      return false;
+    } else {
+      //logger.info("Adding item to inventory - " + item.getComponent(ItemComponent.class).getItemName() + ",old count " + this.itemCount.getOrDefault(item.getComponent(ItemComponent.class).getItemName(), 0));
```
Comment (JTrenerry) [inline source/core/src/main/com/csse3200/game/components/player/InventoryComponent.java:357]: uncomment / remove

File: source/core/src/main/com/csse3200/game/components/player/InventoryComponent.java [hunk 8]
```diff
@@ -186,13 +374,30 @@ public boolean addItem(Entity item) {
...
+            for (int i = 0; i < this.itemPlace.size(); i++) {
+                if (this.itemPlace.get(i) == item.getComponent(ItemComponent.class).getItemName()) {
+                  this.itemPlace.remove(i);
+                  itemChange(i);
+                    break;
+                }
+            }
+        }
+        entity.getEvents().trigger("updateToolbar");
+        //logger.info("Removing item from inventory - " + item.getComponent(ItemComponent.class).getItemName() + ",new count " + this.itemCount.getOrDefault(item.getComponent(ItemComponent.class).getItemName(), 0));
```
Comment (JTrenerry) [inline source/core/src/main/com/csse3200/game/components/player/InventoryComponent.java:398]: uncomment / remove

File: source/core/src/main/com/csse3200/game/components/player/InventoryComponent.java [hunk 9]
```diff
@@ -202,8 +407,11 @@ public boolean removeItem(Entity item) {
    */
   public void setHeldItem(int index) {
     if (index >= 0 && index < 10) {
-      this.heldItem = itemPlace.get(index);
+      this.heldItem = this.heldItemsEntity.get(this.itemPlace.get(index));
       this.heldIndex = index;
+      //Create a new item from the String in itemPlace.get(index)
+//      this.heldItem = new Entity(this.itemPlace.get(index));
+//      this.heldIndex = index;
```
Comment (JTrenerry) [inline source/core/src/main/com/csse3200/game/components/player/InventoryComponent.java:414]: remove?

**Review by JTrenerry — COMMENTED** [review]
added some comments

File: source/core/src/main/com/csse3200/game/components/player/InventoryComponent.java [hunk 10]
```diff
@@ -1,183 +1,383 @@
...
-    for (int i = 0; i < itemPlace.size();i++) {
-      if (itemPlace.getOrDefault(i,null) == null) {
-        return i;
-      }
-      else {
-        if (i == 30) {
-          return -1;
-        }
-      }
+  private HashMap<String,Entity> heldItemsEntity = new HashMap<>();
```
Comment (tristanduncombe) [inline source/core/src/main/com/csse3200/game/components/player/InventoryComponent.java:23]: Could all 4 of these hashmaps not be combined into 1? I don't get why we need 4 different maps where we can have one map containing all the data.

File: source/core/src/main/com/csse3200/game/components/player/InventoryComponent.java [hunk 11]
```diff
@@ -231,76 +451,41 @@ public int getHeldIndex() {
...
-   */
-  public void setItemPosition(Entity item, Point point) {
-    itemPosition.put(item, point);
-  }
+  public void itemChange(int index) {
+    if(index >= 0 && index < 10) {
+      // Only update toolbar
 
+    } else {
+      // update both toolbar and inventory
```
Comment (tristanduncombe) [inline source/core/src/main/com/csse3200/game/components/player/InventoryComponent.java:487]: Does nothing at this point, and you even if the index is less than 10, you still need to update both. Function doesn't make sense or have any tangible purpose.

File: source/core/src/main/com/csse3200/game/components/inventory/ItemSlot.java [hunk 12]
```diff
@@ -72,6 +74,7 @@ public ItemSlot(boolean selected) {
      */
     public void setCount(Integer count) {
         this.count = count;
+        createItemSlot();
```
Comment (tristanduncombe) [inline source/core/src/main/com/csse3200/game/components/inventory/ItemSlot.java:77]: This shouldn't be too bad, I think there might be better approaches but this is nice at simple. Doesn't affect performance

**Review by tristanduncombe — CHANGES_REQUESTED** [review]
Definitely not in a merge-able state, a lot of useless functions which are either not implemented or have a load of commented code within them. Additionally, there are many simplifications that could have been enacted in this ticket, which did not occur.

File: source/core/src/main/com/csse3200/game/components/player/InventoryComponent.java (hunk 11 above)
Comment (nimeshgarg) [reply source/core/src/main/com/csse3200/game/components/player/InventoryComponent.java:487]: The function was to trigger updates of specific itemSlot but I think you took care of it in the display itself.

Removing the function.

**Comment by nimeshgarg** [comment]
Commented line in entity.java - 493,494 for read function 
Read and write functions need to be done for save/load

**Review by JTrenerry — APPROVED** [review]

**Review by tristanduncombe — APPROVED** [review]

**Pull request merged**